### PR TITLE
[SPARK-24809] [SQL] Serializing LongToUnsafeRowMap in executor may result in data error

### PR DESCRIPTION
**Pull request by liutang123**
When join key is long or int in broadcast join, Spark will use `LongToUnsafeRowMap` to store key-values of the table witch will be broadcasted. But, when `LongToUnsafeRowMap` is broadcasted to executors, and it is too big to hold in memory, it will be stored in disk. At that time, because `write` uses a variable `cursor` to determine how many bytes in `page` of `LongToUnsafeRowMap` will be write out and the `cursor` was not restore when deserializing, executor will write out nothing from page into disk.

## What changes were proposed in this pull request?
Restore cursor value when deserializing.

### CODE REVIEW

**Comment by hvanhovell** [comment]
ok to test

**Comment by hvanhovell** [comment]
@liutang123 can you explain why we are losing data when serializing to disk. Also can you add a unit test?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala [hunk 1]
```diff
@@ -726,8 +726,9 @@ private[execution] final class LongToUnsafeRowMap(val mm: TaskMemoryManager, cap
 
     writeLong(array.length)
     writeLongArray(writeBuffer, array, array.length)
-    val used = ((cursor - Platform.LONG_ARRAY_OFFSET) / 8).toInt
-    writeLong(used)
+    val cursorFlag = cursor - Platform.LONG_ARRAY_OFFSET
+    writeLong(cursorFlag)
+    val used = (cursorFlag / 8).toInt
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: Is this said, that when `(cursor - Platform.LONG_ARRAY_OFFSET) / 8` is over the range of Int, we will have overflow? But later you still do `toInt` and use the value?

Comment (liutang123) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: losing data when serializing LongHashedRelation in executor, can you see [this picture](http://oi67.tinypic.com/2z5pzs7.jpg)? In executor, the cursor is 0.

**Comment by liutang123** [comment]
@hvanhovell Thanks for reviewing. Losing data because the variable **cursor** in executor is 0 and serialization depends on it. I will add an UT later.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala (hunk 1 above)
Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: Can you post the image in this PR? The web site you refer contains too many ads.

Comment (liutang123) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: image

**Comment by viirya** [comment]
Let me clarify it. So this means that when `LongToUnsafeRowMap` is broadcasted to executors, and it is too big to hold in memory, it will be stored in disk. At that time, because `write` uses `cursor` to determine `used`, it will write out nothing from `page` into disk.

Is this what you mean?

**Comment by liutang123** [comment]
@viirya Yes, absolutely right. :)

**Comment by liutang123** [comment]
Jenkins test this please

**Comment by jinxing64** [comment]
Jenkins, test this please

**Comment by liutang123** [comment]
@viirya Hi, Could you have more time to review this PR?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala [hunk 2]
```diff
@@ -726,9 +726,10 @@ private[execution] final class LongToUnsafeRowMap(val mm: TaskMemoryManager, cap
 
     writeLong(array.length)
     writeLongArray(writeBuffer, array, array.length)
-    val used = ((cursor - Platform.LONG_ARRAY_OFFSET) / 8).toInt
-    writeLong(used)
-    writeLongArray(writeBuffer, page, used)
+
+    val usedWordsNumber = ((cursor - Platform.LONG_ARRAY_OFFSET) / 8).toInt
+    writeLong(usedWordsNumber)
+    writeLongArray(writeBuffer, page, usedWordsNumber)
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: If no good reason, shall we revert this change? Looks like you only rename it?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala [hunk 3]
```diff
@@ -770,8 +771,10 @@ private[execution] final class LongToUnsafeRowMap(val mm: TaskMemoryManager, cap
     val length = readLong().toInt
     mask = length - 2
     array = readLongArray(readBuffer, length)
-    val pageLength = readLong().toInt
-    page = readLongArray(readBuffer, pageLength)
+    val usedWordsNumber = readLong().toInt
+    // Set cursor because cursor is used in write function.
+    cursor = usedWordsNumber * 8 + Platform.LONG_ARRAY_OFFSET
+    page = readLongArray(readBuffer, usedWordsNumber)
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: ditto. Can you just update `cursor` and revert other unrelated change?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/joins/HashedRelationSuite.scala [hunk 4]
```diff
@@ -278,6 +278,39 @@ class HashedRelationSuite extends SparkFunSuite with SharedSQLContext {
     map.free()
   }
 
+  test("SPARK-24809: Serializing LongHashedRelation in executor may result in data error") {
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/joins/HashedRelationSuite.scala]: Is it possible to have an end-to-end test for this?

**Comment by viirya** [comment]
@liutang123 Thanks for this work. I'm curious that if this is an actual problem you hit in real application, or you just think it is problematic?

**Comment by liutang123** [comment]
@viirya This case  occurred in our cluster and we took a lot of time to find this bug.
For some man-made reasons, the small table's max id has become abnormally large. The LongHasedRelation generated based on the table was not optimized to `dense` and has become abnormally big(approximately 400MB).

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala [hunk 5]
```diff
@@ -772,6 +772,8 @@ private[execution] final class LongToUnsafeRowMap(val mm: TaskMemoryManager, cap
     array = readLongArray(readBuffer, length)
     val pageLength = readLong().toInt
     page = readLongArray(readBuffer, pageLength)
+    // Set cursor because cursor is used in write function.
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: maybe: `Restore cursor variable to make this map able to be serialized again on executors`?

**Comment by viirya** [comment]
As you actually modify `LongToUnsafeRowMap`, is it better to update the PR title and description to reflect that?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/joins/HashedRelationSuite.scala [hunk 6]
```diff
@@ -278,6 +278,39 @@ class HashedRelationSuite extends SparkFunSuite with SharedSQLContext {
...
+    val unsafeProj = UnsafeProjection.create(Array[DataType](LongType))
+    val originalMap = new LongToUnsafeRowMap(mm, 1)
+
+    val key1 = 1L
+    val value1 = new Random().nextLong()
+
+    val key2 = 2L
+    val value2 = new Random().nextLong()
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/joins/HashedRelationSuite.scala]: Is it necessary to use `Random` here? Can we use two arbitrary long values?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/joins/HashedRelationSuite.scala [hunk 7]
```diff
@@ -278,6 +278,39 @@ class HashedRelationSuite extends SparkFunSuite with SharedSQLContext {
...
+
+    originalMap.append(key1, unsafeProj(InternalRow(value1)))
+    originalMap.append(key2, unsafeProj(InternalRow(value2)))
+    originalMap.optimize()
+
+    val resultRow = new UnsafeRow(1)
+    assert(originalMap.getValue(key1, resultRow).getLong(0) === value1)
+    assert(originalMap.getValue(key2, resultRow).getLong(0) === value2)
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/joins/HashedRelationSuite.scala]: We don't need to test `LongToUnsafeRowMap`'s normal feature here. We just need to verify the map after two ser/de can work normally.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/joins/HashedRelationSuite.scala [hunk 8]
```diff
@@ -278,6 +278,39 @@ class HashedRelationSuite extends SparkFunSuite with SharedSQLContext {
...
+
+    val ser = new KryoSerializer(
+            (new SparkConf).set("spark.kryo.referenceTracking", "false")).newInstance()
+
+    val mapSerializedInDriver = ser.deserialize[LongToUnsafeRowMap](ser.serialize(originalMap))
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/joins/HashedRelationSuite.scala]: nit:
```scala
// Simulate serialize/deserialize twice on driver and executor
val firstTimeSerialized = ...
val secondTimeSerialized = ...
```

**Comment by viirya** [comment]
cc @cloud-fan

**Comment by liutang123** [comment]
Jenkins, test this please

**Comment by kiszk** [comment]
retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/execution/joins/HashedRelationSuite.scala [hunk 9]
```diff
@@ -278,6 +278,36 @@ class HashedRelationSuite extends SparkFunSuite with SharedSQLContext {
...
+    val value1 = 4852306286022334418L
+
+    val key2 = 2L
+    val value2 = 8813607448788216010L
+
+    originalMap.append(key1, unsafeProj(InternalRow(value1)))
+    originalMap.append(key2, unsafeProj(InternalRow(value2)))
+    originalMap.optimize()
+
+    val ser = new KryoSerializer(
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/joins/HashedRelationSuite.scala]: we can write `sparkContext.env.serializer.newInstance()`

**Comment by cloud-fan** [comment]
good catch! LGTM

**Comment by viirya** [comment]
LGTM too.

**Review by gatorsmile — COMMENTED** [review]
Thanks! Merged to master/2.3/2.2/2.1
